### DISCUSSION TITLE
Setting ADVectorVariableValues from individual coupled variable components

### DISCUSSION THREAD

**Discussion by rtaylo45** [idaholab/moose#18278] [Q&A General] [answered]
Hello,

I have a coupled 1D  Thermal hydraulics MOOSE app that solves for, among other things, velocity. I would like to use this code in MultiApps to transfer its velocity to my MOOSE app and use this velocity to transport material. But this code returns velocity as a scalar and not a vector. I have written an advection kernel shown below with the header file:

```
class MoleAdvection : public ADKernelValue
{
public:
  static InputParameters validParams();
  MoleAdvection(const InputParameters & parameters);
  virtual ~MoleAdvection();

protected:
  virtual ADReal precomputeQpResidual() override;

  // x velocity
  const ADVariableValue & _x_velocity;
  // y velocity
  const ADVariableValue & _y_velocity;
  // z velocity
  const ADVariableValue & _z_velocity;
  // Velocity vector
  const ADVectorVariableValue & _vector_velocity;
};
```

and the C file

```
#include "MoleAdvection.h"

registerADMooseObject("moleApp", MoleAdvection);

InputParameters
MoleAdvection::validParams()
{
  InputParameters params = ADKernelValue::validParams();
  params.addClassDescription("Applies non-conservative advection");
  params.addCoupledVar("x_velocity", 0., "Coupled x velocity");
  params.addCoupledVar("y_velocity", 0., "Coupled y velocity");
  params.addCoupledVar("z_velocity", 0., "Coupled z velocity");
  params.addCoupledVar("vector_velocity", {0., 0., 0.}, "Coupled vector velocity");
  return params;
}

MoleAdvection::MoleAdvection(const InputParameters & parameters)
  : ADKernelValue(parameters),
  _x_velocity(adCoupledValue("x_velocity")),
  _y_velocity(adCoupledValue("y_velocity")),
  _z_velocity(adCoupledValue("z_velocity")),
  _vector_velocity(adCoupledVectorValue("vector_velocity"))
{
}

MoleAdvection::~MoleAdvection() {}

ADReal
MoleAdvection::precomputeQpResidual()
{
  if (_vector_velocity[_qp].norm() < 1.e-8)
  {
    ADVectorVariableValue _velocity = {_x_velocity, _y_velocity, _z_velocity};
    return _velocity[_qp] * _grad_u[_qp];
  }
  else
  {
    return _vector_velocity[_qp] * _grad_u[_qp];
  }
}
```

 Which does not compile. I have looked at the Doxygen for MOOSE and libMesh to try to figure out how to create a way to create a velocity vector from `x_velocity`, `y_velocity` and `z_velocity`. But i have come up short with trying to figure out how to do it. I would like the kernel to be able to use either a coupled velocity already in vector form (see variable `_vector_velocity` ) or create a velocity vector from the individual components. I am having trouble coming up with a way to do this and could use some help.

**Comment by GiudGiud** [comment, marked as answer]
Hello

To create a velocity vector from individual components, you can use:
ADRealVectorValue V = {_x_velocity[_qp], _y_velocity[_qp], _w_velocity[_qp]};

I dont think you need to instantiate the destructor btw. 

Guillaume

**Reply by rtaylo45** [reply]
I kept having `<brace-enclosed initializer list>` error but this worked:
```
ADRealVectorValue V;
V =  {_x_velocity[_qp], _y_velocity[_qp], _w_velocity[_qp]};
```

**Reply by rtaylo45** [reply]
Actually i need to do this for it to converge.
```
    ADRealVectorValue _velocity(_x_velocity[_qp], _y_velocity[_qp], _z_velocity[_qp]);
    return _velocity * _grad_u[_qp];
```
